astropy/cosmology/flrw/tests/test_base.py

```diff
@@ -1019,8 +1019,6 @@ class HASOde0SubClass(cosmo_cls):
                 def __init__(self, Ode0):
                     pass
 
-            _COSMOLOGY_CLASSES.pop(HASOde0SubClass.__qualname__, None)
```

Comment:
From my testing of putting a print statement before and after the subclass init. I don't believe this line is getting called. It will error on the subclass init and never get to this line. And I can't move it out of the context manager or else we get an UnboundLocalError. I'm not sure the best way to keep this line in.

Reply:
That's an interesting issue. It's important to make sure that `_COSMOLOGY_CLASSES` does not contain `HASOde0SubClass` after the test. Perhaps for now you can put the line in a `try: ... except: UnboundLocalError: ...`?

Reply:
I'm not sure I see how to use a yield fixture here. From reading the documentation it seems like it is used in place of a return. But in this test I dont see a use for a return even from a helper function. The TypeError is raised as soon as the class is declared. We can't return the class definition.

 In fact I don't even think the try block will doing anything. The class is declared in the with statement. We can't reference it anymore so the try will always go to the except block. Not sure the best way to proceed on this one.

Reply:
The `yield` fixture I had in mind would be used with the [`usefixtures` marker](https://docs.pytest.org/en/stable/how-to/fixtures.html#use-fixtures-in-classes-and-modules-with-usefixtures). See the linked documentation for examples.

Reply:
So if I'm understanding it right. Everything before the yield happens before the test then when the test completes everything after the yield runs. So I could deep copy _COSMOLOGY_CLASSES before, then after the test runs set _COSMOLOGY_CLASSES back to the deep copy. Like nothing ever happened. However when I tried this broke other tests, raising Keyerrors in other tests like test_w0cdm.py and test_parameters.py and some others. All keyerrors happen while trying to cleanup by popping a test subclass from _COSMOLOGY_CLASSES. My approach must not be correct or I must be missing something.

Reply:
In astropy/cosmology/flrw/tests/conftest.py
```
import pytest
import copy
from astropy.cosmology import core #imports are where the rest of the imports are, just showing what I imported.

@pytest.fixture
def clean_cosmology_classes():
    original = copy.deepcopy(core._COSMOLOGY_CLASSES)
    yield
    core._COSMOLOGY_CLASSES = original
```

In astropy/cosmology/flrw/tests/test_base.py
```
    @pytest.mark.usefixtures("clean_cosmology_classes")
    def test_init_subclass(self, cosmo_cls):
        """Test initializing subclass, mostly that can't have Ode0 in init."""
        super().test_init_subclass(cosmo_cls)

        with pytest.raises(TypeError, match="subclasses of"):

            class HASOde0SubClass(cosmo_cls):
                def __init__(self, Ode0):
                    pass

```
Test ran `pytest astropy/cosmology`

Reply:
It is best to replace the teardown part of your fixture (i.e. `core._COSMOLOGY_CLASSES = original`) with
```python
core._COSMOLOGY_CLASSES.clear()
core._COSMOLOGY_CLASSES.update(original)
```
The latter avoids assignments, so it is guaranteed to only modify the items in `_COSMOLOGY_CLASSES` and not its identity.

Reply:
Gotcha makes sense, thanks for the help!